### PR TITLE
[GOBBLIN-1071] Retry task initialization

### PR DESCRIPTION
**Pull request by autumnust**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- https://issues.apache.org/jira/browse/GOBBLIN-1071

### Description
The major part of this PR are: 
- In `GobblinMultiTaskAttempt.java`, created a method `createTaskWithRetry`  to wrap around `createTaskRunnable`, so that the task-initialization as part of `createTaskRunnable`  could be retried. The retry-component is built using `Retryer` library which essentially uses guava's retry library. To make that easier, also added some features in `RetryerFactory.java`, especially attempt-based retry to better fit our requirement here. This part is tested within `TaskErrorIntegrationTest.java ` where sits a method `extractorCreationError`. We created a test case for `extractorCreationErrorWithRetry` to demonstrate retry case. 
- In `GobblinHelixTask` also introduced retry of calling `SingleTask`'s constructor in the same approach. The relevant testing is in `GobblinHelixTaskTest.java` where we fail the constructor call on purpose and since we cannot easily heal the failure, it results in a retry-failure at the end. 

This PR also contains many renaming for refactoring as I went through the code base. Most of the rename is because the original name is far from being accurate. 

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java [hunk 1]
```diff
@@ -50,49 +54,74 @@
...
+      TaskAttemptBuilder taskAttemptBuilder, StateStores stateStores, Config dynamicConfig) throws IOException {
     _jobId = jobId;
     _workUnitFilePath = workUnitFilePath;
     _jobStateFilePath = jobStateFilePath;
     _fs = fs;
     _taskAttemptBuilder = taskAttemptBuilder;
     _stateStores = stateStores;
     _dynamicConfig = dynamicConfig;
+
+    _workUnits = getWorkUnits();
```
Comment (sv2000) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java]: Why do we have to eagerly initialize workunits?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java [hunk 2]
```diff
@@ -50,49 +54,74 @@
...
 
     try (SharedResourcesBroker<GobblinScopeTypes> globalBroker = SharedResourcesBrokerFactory
         .createDefaultTopLevelBroker(jobConfig, GobblinScopeTypes.GLOBAL.defaultScopeInstance())) {
-      SharedResourcesBroker<GobblinScopeTypes> jobBroker = getJobBroker(jobState, globalBroker);
+      SharedResourcesBroker<GobblinScopeTypes> jobBroker = getJobBroker(_jobState, globalBroker);
+
+      // Secure atomicity of taskAttempt's execution.
+      _lock.lock();
+      try {
+        _taskAttempt = _taskAttemptBuilder.build(_workUnits.iterator(), _jobId, _jobState, jobBroker);
```
Comment (sv2000) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java]: Looks like the attempt creation can be moved outside the critical section.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java [hunk 3]
```diff
@@ -50,49 +54,74 @@
...
+        _taskAttempt.runAndOptionallyCommitTaskAttempt(GobblinMultiTaskAttempt.CommitPolicy.IMMEDIATE);
+        _taskAttemptBuilt.signal();
```
Comment (sv2000) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java]: How do we handle missed signals? Are we sure the thread waiting on the signal is waiting?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java (hunk 1 above)
Comment (autumnust) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java]: Originally was thinking to have all heavy-lifting work in constructor ( and also make retry more meaning ful there) I don't have strong opinion on this though..

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTask.java [hunk 4]
```diff
@@ -44,6 +49,9 @@
 import org.apache.gobblin.source.workunit.WorkUnit;
 import org.apache.gobblin.util.Id;
 
+import static org.apache.gobblin.util.retry.RetryerFactory.RETRY_INTERVAL_MS;
```
[the remark this answers is not in the document]
Comment (autumnust) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTask.java]: Addressed, thanks for catching.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTask.java [hunk 5]
```diff
@@ -97,28 +105,35 @@ public GobblinHelixTask(TaskRunnerSuiteBase.Builder builder,
...
-                               dynamicConfig);
+    final Config taskLevelConfig = builder.getConfig()
+        .withValue(GobblinClusterConfigurationKeys.TASK_RUNNER_HOST_NAME_KEY, ConfigValueFactory.fromAnyRef(builder.getHostName()))
+        .withValue(GobblinClusterConfigurationKeys.CONTAINER_ID_KEY, ConfigValueFactory.fromAnyRef(builder.getContainerId()))
+        .withValue(GobblinClusterConfigurationKeys.HELIX_INSTANCE_NAME_KEY, ConfigValueFactory.fromAnyRef(builder.getInstanceName()))
+        .withValue(GobblinClusterConfigurationKeys.HELIX_JOB_ID_KEY, ConfigValueFactory.fromAnyRef(this.helixJobId))
+        .withValue(GobblinClusterConfigurationKeys.HELIX_TASK_ID_KEY, ConfigValueFactory.fromAnyRef(this.helixTaskId))
+        .withValue(GobblinClusterConfigurationKeys.HELIX_PARTITION_ID_KEY, ConfigValueFactory.fromAnyRef(partitionNum));
+
+    Retryer<SingleTask> retryer = RetryerFactory.newInstance(taskLevelConfig);
```
[the remark this answers is not in the document]
Comment (autumnust) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixTask.java]: Regarding to the default setting, the timeout is 5 minutes ( it uses TimeUnit.MINUTES.toMillis to convert 5 minutes to ms )

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java [hunk 6]
```diff
@@ -456,6 +464,33 @@ private Task createTaskRunnable(WorkUnitState workUnitState, CountDownLatch coun
...
+    // An "effectively final" variable for counting how many retried has been done, mostly for logging purpose.
+    final AtomicInteger counter = new AtomicInteger(0);
+
+    try {
+      return retryer.call(new Callable<Task>() {
+        @Override
+        public Task call()
+            throws Exception {
+          counter.incrementAndGet();
+          log.info(String.format("The %s time that trying create Task object", counter.get()));
```
Comment (sv2000) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java]: "Task creation Attempt# %s"?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java [hunk 7]
```diff
@@ -456,6 +464,33 @@ private Task createTaskRunnable(WorkUnitState workUnitState, CountDownLatch coun
...
+          return createTaskRunnable(workUnitState, countDownLatch);
+        }
+      });
+    } catch (Exception e) {
+      throw new RuntimeException("Execution in creating a Task-with-retry failed", e);
```
Comment (sv2000) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java]: "Exception creating Task after k retries"?

**Comment by sv2000** [comment]
@autumnust LGTM. Once travis passes, will merge.

**Pull request closed without merging**